docs/rfc/rfc-009-consensus-parameter-upgrades.md

```diff
@@ -0,0 +1,110 @@
+# RFC 009 : Consensus Parameter Upgrade Considerations
+
+## Changelog
+
+- 06-Jan-2011: Initial draft (@williambanfield).
+
+## Abstract
+
+This document discusses the challenges of adding additional consensus parameters
+to Tendermint and proposes a few solutions that can enable addition of consensus
+parameters in a backwards-compatible way.
+
+## Background
+
+This section provides an overview of the issues of adding consensus parameters
+to Tendermint.
+
+### Hash Compatibility
+
+Tendermint produces a hash of a subset of the consensus parameters. The values
+that are hashed currently are the `BlockMaxGas` and the `BlockMaxSize`. These
+are currently in the [HashedParams struct][hashed-params]. This hash is included
+in the block and validators use it to validate that their local view of the consensus
+parameters matches what the rest of the network is configured with.
+
+Any additional consensus parameters added to Tendermint should be included in this
+hash. This presents a challenge for verification of historical blocks when consensus
+parameters are added. If a network produced blocks with a version of Tendermint that
+did not yet have the new consensus parameters, the parameter hash it produced will
+not reference the new parameters. Any nodes joining the network with the newer
+version of Tendermint will have the new consensus parameters. Tendermint will need
+to handle this case so that new versions of Tendermint with new consensus parameters
+can still validate old blocks correctly without having to do anything overly complex
+or hacky. 
+
+### Allowing Developer Defined Values and the `EndBlock` Problem
+
+When new consensus parameters are added, application developers may wish to set
+values for them so that the developer-defined values may be used as soon as the 
+software upgrades. We do not currently have a clean mechanism for handling this.
+
+Consensus parameter updates are communicated from the application to Tendermint
+within `EndBlock` of some height `H` and take effect at the next height, `H+1`.
```

Comment:
Just to check my understanding: The parameters agreed upon affect the consensus rounds that _lead_ to H+1, and it's not that they take effect _after_ H+1 is decided?

Reply:
The application sends updates to these parameters as part of the EndBlock response.

https://github.com/tendermint/tendermint/blob/0ae974e63911804d4a2007bd8a9b3ad81d6d2a90/abci/types/types.pb.go#L2057
That means that the EndBlock response at `H` will affect `H+1`. 

The issue is that height `H` must be decided using some value for the new consensus parameters and that the values for the new consensus parameters cannot be dictated by the application. There's sort of a now-mans-land height here where the defaults would be used. Does that make sense?

Reply:
I think I asked my question poorly. What I mean is, does height H begin or end at the commit for height H? If it "ends" at its commit, is the period during which the validators are negotiating over the contents of future height H+1 affected or not-affected by the parameter changes?

What I think you're saying is "the negotiation period for H+1 uses the old parameters", but I'm not sure why that would be necessary.